kmod/toa/toa.c

```diff
@@ -58,6 +91,177 @@ struct toa_stats_entry toa_stats[] = {
 
 DEFINE_TOA_STAT(struct toa_stat_mib, ext_stats);
 
+
+
+
+#ifdef TOA_IPV6_ENABLE
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3,4,103)
+/* more secured version of ipv6_addr_hash() */
+static inline u32
+__ipv6_addr_jhash(const struct in6_addr *a, const u32 initval)
+{
+	u32 v = (__force u32)a->s6_addr32[0] ^ (__force u32)a->s6_addr32[1];
+
+	return jhash_3words(v,
+			    (__force u32)a->s6_addr32[2],
+			    (__force u32)a->s6_addr32[3],
+			    initval);
+}
+#endif
+
+static void
+toa_ip6_hash(struct toa_ip6_entry *ptr_ip6_entry)
+{
+	struct toa_ip6_data *ptr_toa_data = &ptr_ip6_entry->toa_data;
+	__u32 hash_key =
+		__ipv6_addr_jhash(&ptr_toa_data->in6_addr, ptr_toa_data->port) & TOA_IP6_TAB_MASK;
+
+	spin_lock_bh(&__toa_ip6_list_tab[hash_key].lock);
+
+	list_add(&ptr_ip6_entry->list, &__toa_ip6_list_tab[hash_key].toa_ip6_head);
+
+	spin_unlock_bh(&__toa_ip6_list_tab[hash_key].lock);
+
+	return;
+}
+
+static void
+toa_ip6_unhash(struct toa_ip6_entry *ptr_ip6_entry)
+{
+	struct toa_ip6_data *ptr_toa_data = &ptr_ip6_entry->toa_data;
+	__u32 hash_key =
+		__ipv6_addr_jhash(&ptr_toa_data->in6_addr, ptr_toa_data->port) & TOA_IP6_TAB_MASK;
+
+	spin_lock_bh(&__toa_ip6_list_tab[hash_key].lock);
+
+	list_del(&ptr_ip6_entry->list);
+
+	spin_unlock_bh(&__toa_ip6_list_tab[hash_key].lock);
+}
+
+static void
+lock_all_toa_ip6_sk(void)
+{
+	int i;
+	for_each_possible_cpu(i) {
+		spinlock_t *lock;
+
+		lock = per_cpu_ptr(toa_ip6_sk_lock.lock, i);
+		spin_lock_bh(lock);
+	}
+}
+
+static void
+unlock_all_toa_ip6_sk(void)
+{
+	int i;
+	for_each_possible_cpu(i) {
+		spinlock_t *lock;
+
+		lock = per_cpu_ptr(toa_ip6_sk_lock.lock, i);
+		spin_unlock_bh(lock);
+	}
+}
+
+static void
+lock_cpu_toa_ip6_sk(void)
+{
+	spinlock_t *lock = this_cpu_ptr(toa_ip6_sk_lock.lock);
+	spin_lock_bh(lock);
+}
+
+static void
+unlock_cpu_toa_ip6_sk(void)
+{
+	spinlock_t *lock = this_cpu_ptr(toa_ip6_sk_lock.lock);
+	spin_unlock_bh(lock);
+}
+
+static int
+init_toa_ip6(void)
+{
+	int i;
+
+	for_each_possible_cpu(i) {
+		spinlock_t *lock;
+
+		lock = per_cpu_ptr(toa_ip6_sk_lock.lock, i);
+		spin_lock_init(lock);
+	}
+
+	for (i = 0; i < TOA_IP6_TAB_SIZE; ++i) {
+		INIT_LIST_HEAD(&__toa_ip6_list_tab[i].toa_ip6_head);
+		spin_lock_init(&__toa_ip6_list_tab[i].lock);
+	}
+
+
+	toa_ip6_sk_lock.lock = alloc_percpu(spinlock_t);
+	if (toa_ip6_sk_lock.lock == NULL) {
+		TOA_INFO("fail to alloc per cpu ip6's destruct lock\n");
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int
+exit_toa_ip6(void)
+{
+	int i;
+	struct list_head *head;
+	struct toa_ip6_entry *ptr_ip6_entry;
+	struct sock *sk;
+
+	lock_all_toa_ip6_sk();
+
+	for (i = 0; i < TOA_IP6_TAB_SIZE; ++i) {
+
+		spin_lock_bh(&__toa_ip6_list_tab[i].lock);
```

Comment:
lock_all_toa_ip6_sk will block all cpu in the server， and this function also need wait all the cpu release per cpu lock.  The for {....} will relase all the ip6_toa data,  surely, it will spend times in exit_toa_ip6(), related to ip6_toa data and net's task .   But this function only is called when unload toa.